include/boost/url/impl/url.ipp

```diff
@@ -2042,26 +2025,157 @@ resolve(
 //
 //------------------------------------------------
 
-url&
+void
 url::
-normalize()
+normalize_octets_impl(
+    int id,
+    grammar::lut_chars const& cs) noexcept
 {
-#if 0
-    // scheme
+    char* it = s_ + offset(id);
+    char* end = s_ + offset(id + 1);
+    char buf = 0;
+    char* dest = it;
+    while (it < end)
     {
-        char* it = s_
-            + offset(id_scheme);
-        auto last = it + offset(id_user);
-        if(it < last)
+        if (*it != '%')
+        {
+            *dest = *it;
+            ++it;
+            ++dest;
+            continue;
+        }
+        if (end - it < 3)
+            break;
+
+        // decode unreserved octets
+        pct_decode_unchecked(
+            &buf,
+            &buf + 1,
+            string_view(it, 3));
+        if (cs(buf))
         {
-            --last;
-            while(it != last)
+            *dest = buf;
+            it += 3;
+            ++dest;
+            continue;
+        }
+
+        // uppercase percent-encoding triplets
+        ++it;
+        *it = grammar::ascii_toupper(*it);
+        ++it;
+        *it = grammar::ascii_toupper(*it);
+        ++it;
+        dest += 3;
+    }
+    if (it != dest)
+    {
+        std::size_t diff = it - dest;
+        std::size_t s = len(id) - diff;
+        shrink_impl(id, s);
+        s_[size()] = '\0';
+    }
+}
+
+void
+url::
+to_lower_impl(int id) noexcept
+{
+    char* it = s_ + offset(id);
+    char const* const end = s_ + offset(id + 1);
+    if(it < end)
+    {
+        while(it < end)
+        {
+            if (*it != '%')
             {
-                // to upper
+                *it = grammar::ascii_tolower(
+                    *it);
+                ++it;
+                continue;
             }
+            it += 3;
         }
     }
-#endif
+}
+
+url&
+url::
+normalize_scheme()
+{
+    to_lower_impl(id_scheme);
+    return *this;
+}
+
+url&
+url::
+normalize_authority()
+{
+    // normalize host
+    if (host_type() == urls::host_type::name)
+    {
+        normalize_octets_impl(
+            id_host,
+            detail::reg_name_chars);
+    }
+    to_lower_impl(id_host);
+
+    // normalize password
+    normalize_octets_impl(id_pass, detail::password_chars);
```

Comment:
password holds the leading colon, right? we are giving that character to `normalize_octets_impl`

Reply:
> password holds the leading colon, right?

I think so. 

> we are giving that character to normalize_octets_impl

That's fine. `normalize_octets_impl` doesn't encode stuff.